plugins/csp-simple-bodies/src/SimpleBody.cpp

```diff
@@ -152,9 +159,66 @@ float orenNayar(vec3 N, vec3 L, vec3 V) {
   
   return max(0, (L1 + L2) * cos_theta_i);
 }
-    
-void main()
-{
+
+// Calculates the shading by planetary rings. This is done in 3 steps:
+// 1. Calculate the intersection between a ray from the fragment to the Sun and the ring plane.
+// 2. If an intersection exists check if it falls within the ring.
+// 3. If it falls within the ring, we get the brightness from the ring texture.
+//
+//                 o o o                   \ /
+//             o           o             -- O --
+//           o               o             / \
+// -------  o                 o  -------
+//           o               o
+//             o           o
+//                 o o o
+//
+float getRingShadow() {
+  #ifdef HAS_RING
+    // The up direction of the ring plane.
+    vec3 ringNormal = normalize(vNorth);
+
+    // The normalized direction of the sun.
+    vec3 sunNormal = normalize(vSunDirection);
+
+    // If the ray is parallel to the ring, we don't draw a shadow.
+    float sunAngle = dot(ringNormal, sunNormal);
+    if (sunAngle == 0) {
+      return 1.0;
+    }
+
+    // The distance along the ray from the fragment to the intersection.
+    float t = (dot(ringNormal, vCenter) - dot(ringNormal, vPosition)) / dot(ringNormal, sunNormal);
+
+    // If the distance is negative, the ray intersects the ring away from the Sun.
+    if (t < 0.0) {
+      return 1.0;
+    }
+
+    // The exact point of intersection with the ring plane.
+    vec3 intersect = vPosition + (sunNormal * t);
+
+    // The distance from the bodies center. If it is outside the ring radii, we don't have a shadow.
+    float dist = length(vCenter - intersect);
+    if (dist < uRingRadii.x || dist > uRingRadii.y) {
+      return 1.0;
+    }
+
+    // Convert the distance to the texture coordinate.
+    float texPosition = (dist - uRingRadii.x) / (uRingRadii.y - uRingRadii.x);
+
+    // We apply the same shading logic as in the rings plugin.
+    #ifdef ENABLE_HDR
+      return (1.0 - texture(uRingTexture, vec2(texPosition, 0.5)).a) * 0.5;
+    #else
+      return (1.0 - texture(uRingTexture, vec2(texPosition, 0.5)).a) * 2.0;
+    #endif
```

Comment:
I think this brightness variation is not required. In fact, it leads to weird artifacts especially in non-HDR mode. Also, in the original ring shader this is not applied to the alpha channel. I think this looks better:

```suggestion
    float texPosition = (dist - uRingRadii.x) / (uRingRadii.y - uRingRadii.x);
    return 1.0 - texture(uRingTexture, vec2(texPosition, 0.5)).a;
```